x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java

```diff
@@ -992,9 +993,14 @@ public AsyncSender interceptSender(AsyncSender sender) {
     @Override
     public List<ExecutorBuilder<?>> getExecutorBuilders(final Settings settings) {
         if (enabled) {
-            return Collections.singletonList(
-                    new FixedExecutorBuilder(settings, TokenService.THREAD_POOL_NAME, 1, 1000, "xpack.security.authc.token.thread_pool",
-                        false));
+            final int allocatedProcessors = EsExecutors.allocatedProcessors(settings);
+            return List.of(
+                new FixedExecutorBuilder(settings, TokenService.THREAD_POOL_NAME, 1, 1000,
+                    "xpack.security.authc.token.thread_pool", false),
+                new FixedExecutorBuilder(settings, ApiKeyService.THREAD_POOL_NAME,
+                    (allocatedProcessors + 1) / 2, 1000,
```

Comment:
Note by using half of the allocated processors, initial authentiation of API keys, i.e. warming up the cache, could be up to twice as slow as current implemention.

Reply:
The **maximum theoretical throughput** of new (previously unseen) API key validations will indeed halve, but I'm not personally worried about it. We're talking about the theoretical maximum throughput which, because of the contention on the get threadpool, and because there are many thread pools overbooking the available processors, it's not something I would consider practically important.

Yet, a possible mitigation would be to decrease (halve) the default hashing cost factor for API Keys. GIven the length of the random api key secret that we generate, we would still remain out of the brute force plausibility range.

Reply:
That's quite a large number of new threads to be adding to the system, especially threads that could end up all being busy at once for a long time authenticating a lot of clients. They're then stealing CPU time from other components of the system. What work have we done to justify that the thread pool needs to be this large (the ceiling of half the number of cores)?

Reply:
A few reasons:
* It has fewer threads than the `GET` thread pool - the current code performs the hashing in the `GET` thread pool, which has number of threads equals to number of cores. The new thread pool should in fact help with resource stealing. Besides, saturation of GET thread pool has wider negative impacts to both uncached and cached authentication. A new pool helps isolate the impact to only new clients trying to connect.
* Balance between throughput and resource contention - the expensive hashing operations are unavoidable. Performance tests show about 5000 authentications per minute (8 cores). With half of the threads, the number will be halved as well. Assuming a 25K clients (50K keys) use case, it will take 20 min for all keys to be cached. Reducing the number of threads too much would push the warm up time even higher. Therefore using half number of the cores feels like a reasonable middle ground.
* Hashing in the `generic` thread pool (which uses all cores) does not have obvious performance hit - I tried simply pushing the hashing to the `generic` thread pool, which then used all cores for the hashing. I didn't observe any obivious performance issue with it. In fact, the initial warm up stage is smoother and has a better throughput than using the `GET` thread pool. Therefore it feels that using half of the cores should be relatively safe and also performant. I do plan to run a few more rounds of performance tests after all scheduled `v7.9` API key improvements. So we can be confident on this.
* The new thread pool will be shared by all security related expensive operations - currently its only usage is for hashing of API key auth, so it is not a very good reason. But it could soon see another use case: the API key creation also requires expensive hashing which currently runs in the `transport_worker` thread. I need to do some more investigation, but the result is very likely that we need move it to this new thread pool as well.

Reply:
My _thoughts_  for approving the PR were that we operate under some desired rough approximation for the rate of authentication using keys (what Yang is describing on the second point above) and so by switching to a new pool (given that the get operation is so much faster than the hash, and that the new pool is smaller than the get pool) all the threads in the new pool will be busy at peak rate, so that we can estimate that peak rate (assuming a relatively tranquil system where the pool threads stay running). The original peak rate (using the get pool) was noisy because any get requests in the system interfered with it (in addition to the obvious flip side that authn hashing interfered with get requests in the system).

When the system is busy and all threads are working at the pool queues and no queue is empty, it all boils down to the processors' queue. In a sense all thread pool queues "merge" into processor queues, with the priority given by the relative number of threads ascribed to each pool. So introducing a new pool will decrease the throughput of existing pools, proportionate to relative sizes, in exchange of a guaranteed minimum throughput on the new pool ,because the new queue is not populated with any other types of work (compared to reusing an existing pool that contains other work types).

To summarise, I would generally introduce a thread pool when I want more control over the rate of some work (both the minimum and the maximum rate).

That's my thought process about thread pool queues upon which I've approved the PR. I believe a new thread pool is the right decision, but I concede that we've informed the sizing decision on rough external authn rates without also consulting with folks that tune the existing thread pools.

Reply:
Okay, thanks for the great explanations to help me understand the perspective here. One more question: is it enough to have a simple "security" thread pool? Do we need a token and a crypto thread pool and maybe another thread pool for other security aspects in the future?

Reply:
**TL;DR**: We are not considering another thread pool for other security aspects. 

Our intention for the new thread pool is to focus on "cpu-intensive" security computations, namely hashing and encryption, especially in the scope of authentication. Given their expensiveness and importance, we'd like to control their impact both ways: they should not take over all system resources but in the mean time should maintain a reasonable level of throughput. We use following criteria to justify whether an operation should be added to this pool:
* _Is it expensive_? The standard here is somewhat subjective. Based on performance tests, I personally would eye on anything longer than 1 ms.
* _Is it authentication/authorization related_? We'd like to prioritize these operations since they usually need to complete in a timely manner. 
* _Does it need to scale? i.e. potential burst in volume_? When there is a large increase in concurrency, we'd like to protect the system from being completely flooded. 

With above criteria, we can come up with the followings:
* Hashing for uncached results are great candidates. These include API Key auth (current PR, v7.9), API Key creation (v7.9?), tokens, username/password (planned, v7.10) 
* Signing and cryption for SAML assertions and odic claims are potential candidates (pending further performance check)
* Hashing for cached result is _not_ qualified for the new thread pool since it is fast (sha256, ~3 microseconds). 
* Manipulations of security documents are _not_ qualified either
* Encrypted snapshot, though potentially expensive, is not auth related and unlikely needs to scale. Hence it is also _not_ qualified.
* An interesting case is read/write of the security index. When the index is not on the local node, these operations can be expensive (above 1 ms) and it also matches the two other criteria. However the system indices work (#57936) is supposed to cover it. Hence we are _not_ considering it either.
* Any other security aspects that are _not_ qualified

Now the question is: _whether above _unqualified_ security operations worth to have their own separate thread pool?_

When security is enabled, security related operations are ubiquitous, e.g. every request needs to go through authentication and authorization flow. Currently our *general practice* is to "just let them run in the same thread", which could be `get`, `transport_worker` etc. This strategy has served us reasonally well so far. But with the upcoming change to expose the service to unprecedented large number of clients, we started noticing some issues revealed by the performance tests, which in turn leads to this PR and a few other work in the pipeline. At this stage, based on what we know, we still believe the existing strategy works fine for security operations that are _not_ qualified for the new thread pool. Hence we do _not_ plan to add another one. 

For the sake of completeness, there are exceptions to the existing strategy. Outbound LDAP calls are [submitted via the generic pool](https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ldap/support/LdapUtils.java#L95-L97) to avoid risks of deadlocking if they are submitted via the same pool used by the outbound requests. There is no sign for this being an issue so far. We had brief discussion about it and decided to evaluate it more closely at a later time.